x/gov/client/cli/prompt.go

```diff
@@ -0,0 +1,296 @@
+			prompt.Validate = client.ValidatePromptURL
+		}
+
+		if strings.EqualFold(fieldName, "authority") {
+			// pre-fill with gov address
+			prompt.Default = authtypes.NewModuleAddress(types.ModuleName).String()
+			prompt.Validate = client.ValidatePromptAddress
+		}
+
+		if strings.Contains(fieldName, "addr") ||
+			strings.Contains(fieldName, "sender") ||
+			strings.Contains(fieldName, "voter") ||
+			strings.Contains(fieldName, "depositor") ||
+			strings.Contains(fieldName, "granter") ||
+			strings.Contains(fieldName, "grantee") ||
+			strings.Contains(fieldName, "recipient") {
+			prompt.Validate = client.ValidatePromptAddress
+		}
+
+		result, err := prompt.Run()
+		if err != nil {
+			return data, fmt.Errorf("failed to prompt for %s: %w", fieldName, err)
+		}
+
+		switch v.Field(i).Kind() {
+		case reflect.String:
+			v.Field(i).SetString(result)
+		case reflect.Int:
+			resultInt, _ := strconv.Atoi(result)
+			v.Field(i).SetInt(int64(resultInt))
+		default:
+			// skip other types
+			// possibly in the future we can add more types (like slices)
+			continue
+		}
+	}
+
+	return data, nil
+}
+
+type proposalTypes struct {
+	Type    string
+	MsgType string
+	Msg     sdk.Msg
+}
+
+// Prompt the proposal type values and return the proposal and its metadata
+func (p *proposalTypes) Prompt(cdc codec.Codec) (*proposal, ProposalMetadata, error) {
+	proposal := &proposal{}
+
+	// set metadata
+	metadata, err := Prompt(ProposalMetadata{}, "proposal")
+	if err != nil {
+		return nil, metadata, fmt.Errorf("failed to set proposal metadata: %w", err)
+	}
+
+	proposal.Metadata = fmt.Sprintf("Insert here %s IPFS CID", draftMetadataFileName)
+
+	// set deposit
+	depositPrompt := promptui.Prompt{
+		Label:    "Enter proposal deposit",
+		Validate: client.ValidatePromptCoins,
+	}
+	proposal.Deposit, err = depositPrompt.Run()
+	if err != nil {
+		return nil, metadata, fmt.Errorf("failed to set proposal deposit: %w", err)
+	}
+
+	if p.Msg == nil {
+		return proposal, metadata, nil
+	}
+
+	// set messages field
+	result, err := Prompt(p.Msg, "msg")
+	if err != nil {
+		return nil, metadata, fmt.Errorf("failed to set proposal message: %w", err)
+	}
+
+	message, err := cdc.MarshalInterfaceJSON(result)
+	if err != nil {
+		return nil, metadata, fmt.Errorf("failed to marshal proposal message: %w", err)
+	}
+	proposal.Messages = append(proposal.Messages, message)
+	return proposal, metadata, nil
+}
+
+var supportedProposalTypes = []proposalTypes{
+	{
+		Type:    proposalText,
+		MsgType: "", // no message for text proposal
+	},
+	{
+		Type:    "community-pool-spend",
+		MsgType: "/cosmos.distribution.v1beta1.MsgCommunityPoolSpend",
+	},
+	{
+		Type:    "software-upgrade",
+		MsgType: "/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade",
+	},
+	{
+		Type:    "cancel-software-upgrade",
+		MsgType: "/cosmos.upgrade.v1beta1.MsgCancelUpgrade",
+	},
+	{
+		Type:    proposalOther,
+		MsgType: "", // user will input the message type
+	},
+}
```

Comment:
how do we add other message types here? maybe there is some sort of grpc reflection we can use to grab more message types?

Reply:
If the user select `other` he can type the type him/herself:

![image](https://user-images.githubusercontent.com/29894366/190680608-0e298186-fd95-417c-bcb9-2e0b77c0fa83.png)
![image](https://user-images.githubusercontent.com/29894366/190680845-7f7d8a31-3f16-4cbd-8a06-a41a74fc7d4a.png)

Reply:
nice, do we have a place where users can find this?

I was thinking we automate the part of listing them, we could reuse grpc for this

Reply:
I don't think so, let me check how to have a select prompt instead then 🤔

Reply:
Working: https://asciinema.org/a/c8LWtciM6iX5pYgdISLaYJ7FX